src/ast/passes/macro_expansion.cpp

```diff
@@ -0,0 +1,320 @@
+                const std::unordered_map<std::string, Macro *> &macros,
+                std::vector<std::string> &&macro_stack);
+
+  using Visitor<MacroExpander>::visit;
+
+  void visit(AssignVarStatement &assignment);
+  void visit(Variable &var);
+  void visit(Map &map);
+  void visit(Expression &expr);
+
+  std::optional<Block *> expand(Macro &macro, const Call &call);
+  bool check_recursive_call(const std::string &macro_name, const Call &call);
+
+private:
+  ASTContext &ast_;
+  const std::string macro_name_;
+  const std::unordered_map<std::string, Macro *> &macros_;
+
+  std::string get_new_var_prefix();
+  std::string get_new_var_ident(std::string original_ident);
+
+  // Maps of macro map/var names -> callsite map/var names
+  std::unordered_map<std::string, std::string> maps_;
+  std::unordered_map<std::string, std::string> vars_;
+  std::unordered_set<std::string> temp_vars_;
+  const std::vector<std::string> macro_stack_;
+};
+
+class MacroCollection : public Visitor<MacroCollection> {
+public:
+  MacroCollection(ASTContext &ast, BPFtrace &b);
+  void collect_macros();
+
+  using Visitor<MacroCollection>::visit;
+
+  void visit(Expression &expr);
+
+private:
+  ASTContext &ast_;
+  BPFtrace &bpftrace_;
+  std::unordered_map<std::string, Macro *> macros_;
+};
+
+MacroExpander::MacroExpander(
+    ASTContext &ast,
+    std::string macro_name,
+    const std::unordered_map<std::string, Macro *> &macros,
+    std::vector<std::string> &&macro_stack)
+    : ast_(ast),
+      macro_name_(std::move(macro_name)),
+      macros_(macros),
+      macro_stack_(std::move(macro_stack))
+{
+  assert(!macro_stack_.empty());
+}
+
+void MacroExpander::visit(AssignVarStatement &assignment)
+{
+  auto *var = assignment.var();
+  if (var->ident.starts_with(get_new_var_prefix())) {
+    // Don't change variable names if this is a variable assignment
+    // that we created to represent the passed in expression e.g.
+    // macro add_one($x) { $x + 1 } BEGIN { $a = 1; print(add_one($a + 1)); }
+    // will become
+    // BEGIN { $a = 1; print({let $$add_one_$a = $a + 1; $$add_one_$a + 1}); }
+    return;
+  }
+  if (std::holds_alternative<VarDeclStatement *>(assignment.var_decl)) {
+    visit(std::get<VarDeclStatement *>(assignment.var_decl));
+  } else {
+    visit(std::get<Variable *>(assignment.var_decl));
+  }
+  visit(assignment.expr);
+}
+
+void MacroExpander::visit(Variable &var)
+{
+  if (auto it = vars_.find(var.ident); it != vars_.end()) {
+    var.ident = it->second;
+  } else if (!temp_vars_.contains(var.ident)) {
+    var.ident = get_new_var_ident(var.ident);
+  }
+}
+
+void MacroExpander::visit(Map &map)
+{
+  if (auto it = maps_.find(map.ident); it != maps_.end()) {
+    map.ident = it->second;
+  } else {
+    map.addError() << "Unhygienic access to map: " << map.ident
+                   << ". Maps must be passed into the macro as arguments.";
+  }
+}
+
+bool MacroExpander::check_recursive_call(const std::string &macro_name,
```

Comment:
Lemme think on this and see if I can break it. But I think this approach is correct

Reply:
I think it's correct